kubeadm: Add kubelet instance configuration to configure CRI socket for each node (#4658)
kubernetes/enhancements: keps/sig-cluster-lifecycle/kubeadm/4656-add-instance-configuration/README.md

## Summary

The proposal is to remove the Container Runtime Interface (CRI) socket annotation from the Node object in Kubernetes, which is currently added during the "init upload-config" phase in Kubeadm. This annotation is used to specify the CRI socket endpoint used by the kubelet on each node for communication with the container runtime. 

Instead of relying on this annotation, the proposal recommends not relying on this annotation, but instead creating an instance config per node and overriding the `ContainerRuntimeEndpoint` in the global kubelet config when kubadm init and join. This will eliminate the need for kubeadm to store CRI socket configuration on each node, instead relying on the Kubernetes configuration file to specify this information.

## Motivation

Currently, kubeadm adds a CRI socket annotation to the Node object during the "init upload-config" phase, which specifies the endpoint for the CRI that is being used by the kubelet on each node. This annotation is persistent on the Node object, even if the kubelet is updated or the CRI is changed.

After migrating the container runtime endpoint flag to the instance configuration, we can use it

> Set the cri socket in the global kubelet configuration by overriding the `ContainerRuntimeEndpoint` field in `kubelet-config.yml`.

Reviewer: ```suggestion
Set the CRI socket in the global kubelet configuration by overriding the `ContainerRuntimeEndpoint` field in `kubelet-config.yml`.
```

also currently the node kubelet config file is just called `/var/lib/kubelet/config.yaml`

### Goals

> * Remove the use of CRI socket annotation on Node object.

Reviewer: maybe: Deprecate and remove

> * We will prioritize respecting the local setting over the global one.

Reviewer: this can be removed

> * Add `kubelet-instace-config.yml` and use it to override the `ContainerRuntimeEndpoint` field in `kubelet-config.yml`.

Reviewer: ```suggestion
* Add `kubelet-instance-config.yaml` and use it to override the `ContainerRuntimeEndpoint` field in `kubelet-config.yaml`.
```

extension should be .yaml everywhere in the doc.

should we use instance-config.yaml given the existing file is just config.yaml?

Reviewer: what is instance-config.yaml is it a patch file., similarly to the patch files we use with --patches?
how is kubeadm going to process that file?

> 

Reviewer: seems like this KEP is proposing  kubeadm-flags.env to no longer have --container-runtime-endpoint
this should be in goals.

### Non-Goals

> Update the CRI socket annotation on Node object to be the latest.

Reviewer: i don't think this is a non-goal.

you could add
- Continue maintaining CRI socket paths on Node objects

## Proposal

kubeadm init: If the -cri-socket args is specified, the `-cri-socket` args is preferred and the `kubelet-instace-config.yml` configuration file is generated based on the `-cri-socket` args, if the `-cri-socket` args is not specified, the container runtime ennpoint is automatically detect and the configuration file is uploaded to the global configuration.

kubeadm join:  If the `-cri-socket` args is specified, the `-cri-socket` args is used in preference, and the `kubelet-instace-config.yml `configuration file is generated based on `-cri-socket`, overriding the kubelet configuration downloaded from the global configuration; if the `-cri- socket` args is not specified, the kubelet configuration downloaded from the global configuration will be used.

> 

Reviewer: IMO we should not talk about the kubeadm CLI flag --cri-socket it's an option that ends up in the kubeadm configuration. you should be talking about the "CRI socket supplied in kubeadm configuration"

kubeadm upgrade: `kubeadm upgrade node` will check `kubelet-config-instance.yaml`  only.

### Risks and Mitigations

> ## Design Details

Reviewer: i must admit i don't understand the plan very well. can you please break it down in versions, instead of commands

1.31
- kubeadm init does ....
- kubeadm join ...
- ...

1.32
...

1.33

Author: Do you think it is simple and clear enough now :)

kubeadm join:

* No longer need to add the kubeadm.alpha.kubernetes.io/cri-socket annotation.
* If the `-cri-socket` args is specified, the `-cri-socket` args is used in preference, and the `kubelet-instace-config.yml `configuration file is generated based on `-cri-socket`, overriding the kubelet configuration downloaded from the global configuration; if the `-cri- socket` args is not specified, the kubelet configuration downloaded from the global configuration will be used.

kubeadm reset:

* There is no need to do anything, as kubelet-instace-config.yml will be cleaned up with the deletion of ``/var/lib/kubelet.`

kubeadm upgrade apply:

* In v1.31-1.33, when `kubeadm upgrade apply`, we will read the cri annotaion (if there is no annotaion, we will detect container runtime endpoint automatically) and patch it to the global configuration and generate `kubelet-instace-config`. 
* In v1.34+, `kubeadm upgrade apply` will no longer read cri annotation any more.

kubeadm upgrade node:

> * in v1.31, for other nodes, `kubeadm upgrade node` will check if the cri annotation is diffent with the global setting. if cri-socket is different, We automatically detect the container runtime endpoint to generate `kubelet-instace-config`, and override the `ContainerRuntimeEndpoint` field to `kubelet-config`, but it will not be saved to the global configuration.

Reviewer: i think the plan should be the following instead:
- 1.31 kubeadm upgrade takes the annotations values and starts writing them to instance config.
- 1.32 kubeadm upgrade removes the annotation on node objects.
- 1.32 kubeadm upgrade does nothing.

* `cmd/kubeadm/app/cmd/phases/join`

##### Integration tests

- N/A

##### e2e tests

* A new e2e test will be added by using the kinder tool.

### Graduation Criteria

#### Alpha

- The upgrade will still respect the CRI annotation.
- Add corresponding e2e tests.

#### Beta

- Use instance kubelet configuration or global configuration, ignore CRI annotations.
- Gather feedback from developers and surveys.

#### GA

- Remove the CRI annotation during upgrade(this may be not urgent or have to).

### Upgrade / Downgrade Strategy

During upgrade
We will add the `kubelet-instance-config.yml` file in the `/var/lib/kubelet` directory, define the value of ContainerRuntimeEndpoint in it, and override the `ContainerRuntimeEndpoint` field in the global kubelet configuration.

Downgrade are not supported by kubeadm.

### Version Skew Strategy

kubeadm will continue to skew from kubelet for three versions. The `ContainerRuntimeEndpoint` field in `KubeletConfiguration` was [introduced](https://github.com/kubernetes/kubernetes/pull/112136) in v1.27, so when we overwrite the `ContainerRuntimeEndpoint` field in kubelet-config through instance-config, it will not have any impact on kubelet.

> ## Production Readiness Review Questionnaire

Reviewer: remove the entire section (PRR)

### Dependencies

###### Does this feature depend on any specific services running in the cluster?

### Scalability

###### Will enabling / using this feature result in any new API calls?

No

###### Will enabling / using this feature result in introducing new API types?

No

###### Will enabling / using this feature result in any new calls to the cloud provider?

No

###### Will enabling / using this feature result in increasing size or count of the existing API objects?

No

###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?

No

###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?

No

###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?

No

### Troubleshooting

###### How does this feature react if the API server and/or etcd is unavailable?

###### What are other known failure modes?

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

- \- 2024-05-23: Initial draft KEP

## Drawbacks

> ## Alternatives

Reviewer: how about using a feature gate, and why is that not our first choice?

Author: I initially thought that this feature would be invisible to ordinary users, and feature gate would increase its complexity (and I'm not sure whether it makes sense to use feature gate on kubeadm), so I implemented it in multiple versions to control version skew and compatibility.

If we use feature gate, then the `Design Details` will become relatively simple. We only need to implement the feature and use feature gate to control whether it is enabled. There is no need to use multiple versions implemented to consider compatibility and version skew.

Let me think and redesign it based on feature gate.

Reviewer: if the feature gate is the main approach what alternatives do we have? this section "alternatives" must include one or more alternatives that are not the primary choice of the KEP.

  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

> The proposal is to remove the Container Runtime Interface (CRI) socket annotation from the Node object in Kubernetes, which is currently added during the `init upload-config` phase in Kubeadm. This annotation is used to specify the CRI socket endpoint used by the kubelet on each node for communication with the container runtime. 

Reviewer: add the name of the annotation here.

> Instead of relying on this annotation, the proposal recommends not relying on this annotation, but instead creating an instance config per node and overriding the `ContainerRuntimeEndpoint` in the kubelet config when calling kubeadm init and join. This will eliminate the need for kubeadm to store CRI socket configuration on each node, instead relying on the Kubernetes configuration file to specify this information.

Reviewer: instead of relying on the annotation, this KEP proses ...

on each node -> on each Node object

>  instead relying on the Kubernetes configuration file to specify this information.

you mean a kubeadm configuration file?

## Motivation

Currently, kubeadm adds a CRI socket annotation to the Node object during the `init upload-config` phase, which specifies the endpoint for the CRI that is being used by the kubelet on each node. This annotation is persistent on the Node object, even if the kubelet is updated or the CRI is changed.

After migrating the container runtime endpoint flag to the instance configuration, we can use it
Set the CRI socket by overriding the `ContainerRuntimeEndpoint` field in `/var/lib/kubelet/config.yaml`.

### Goals

* kubeadm currently adds an annotation with the key `kubeadm.alpha.kubernetes.io/cri-socket` to each Node object. We will deprecate and remove it.

> * Add `/var/lib/kubelet/instace-config.yaml`, define the `ContainerRuntimeEndpoint` field in it, and use it to override the `ContainerRuntimeEndpoint` field in `/var/lib/kubelet/config.yaml`.

Reviewer 2: instace --> instance?

> * Add `/var/lib/kubelet/instace-config.yaml`, define the `ContainerRuntimeEndpoint` field in it, and use it to override the `ContainerRuntimeEndpoint` field in `/var/lib/kubelet/config.yaml`.

Reviewer: is `/var/lib/kubelet/instace-config.yaml` a patch file, what is it? how is it applied on top of `.../config.yaml`?

* The `--container-runtime-endpoint` flag is no longer written to the `/var/lib/kubelet/kubeadm-flags.env` file

### Non-Goals

- Continue maintaining CRI socket paths on Node objects.

## Proposal

> kubeadm init: If the CRI socket provided in the kubeadm configuration is set, it will take precedence and generate the `/var/lib/kubelet/instace-config.yaml` configuration file based on it; if the CRI socket is not specified, the container runtime ennpoint will be automatically detected and the configuration file will be uploaded to the global configuration.

Reviewer: ennpoint->endpoint

so if the kubeadm configuration criSocket field is set we write an instance-config.yaml?
what if the criSocket is auto-detected, would we also write an instance-config.yaml in that situation too?

i think, yes. which means the instance-config.yaml is always created, right?

> kubeadm join: If the CRI socket provided in the kubeadm configuration is set, it will take precedence and generate the `/var/lib/kubelet/instace-config.yaml` configuration file based on it, overwriting the kubelet configuration downloaded from the global configuration; if the CRI socket is not specified, the kubelet configuration downloaded from the global configuration is used, and /var/lib/kubelet/instace-config.yaml is generated based on it.

Reviewer: same Q as above.
put /var/lib/kubelet/instace-config.yaml  in quotes.

> kubeadm upgrade: `kubeadm upgrade node` will only check `/var/lib/kubelet/instace-config.yaml`.

Reviewer: kubeadm upgrade: future versions of `kubeadm upgrade apply/node`...

### Risks and Mitigations

## Design Details

> **Alpha(v.1.31):** 

Reviewer 2: ```suggestion
**Alpha(v1.31):** 
```

kubeadm init:

> * No longer need to write the `container-runtime-endpoint` to `/var/lib/kubelet/kubeadm-flags.env`.

Reviewer: --container-runtime-endpoint

* No longer need to add the `kubeadm.alpha.kubernetes.io/cri-socket` annotation.

> * If the CRI socket provided in the kubeadm configuration is set, it is used first and the `/var/lib/kubelet/instace-config.yaml` configuration file is generated based on it. If the CRI socket is not set, the container runtime endpoint is automatically detected and the configuration file is uploaded to the global configuration.

Reviewer: again, what if it's not set in the kubeadm config but it's autodetected?

kubeadm join:

> * No longer need to add the kubeadm.alpha.kubernetes.io/cri-socket annotation.

Reviewer: add quotes

* If the CRI socket provided in the kubeadm configuration is set, it will take precedence and generate the `/var/lib/kubelet/instace-config.yaml` configuration file based on it, overwriting the kubelet configuration downloaded from the global configuration; if the CRI socket is not set, the kubelet configuration downloaded from the global configuration is used, and /var/lib/kubelet/instace-config.yaml is generated based on it.

kubeadm reset:

> * There is no need to do anything, as `/var/lib/kubelet/instace-config.yaml` will be cleaned up with the deletion of ``/var/lib/kubelet.`

Reviewer 2: ```suggestion
* There is no need to do anything, as `/var/lib/kubelet/instace-config.yaml` will be cleaned up with the deletion of `/var/lib/kubelet.`
```

kubeadm upgrade apply:

* `kubeadm upgrade apply` will read the CRI annotaion and patch it to the global configuration and generate `/var/lib/kubelet/instace-config.yaml`, and override the `ContainerRuntimeEndpoint` field to `/var/lib/kubelet/config.yaml`.

kubeadm upgrade node:

> * for other nodes, `kubeadm upgrade node` will check if the CRI annotation and  generate `/var/lib/kubelet/instace-config.yaml` based on them, and override the `ContainerRuntimeEndpoint` field to `/var/lib/kubelet/config.yaml`.

Reviewer: can it check the kubeadm-flags.env instance to avoid an API call?

Author: Yes, we can do that.

**Beta(v1.32)**:

> * `kubeadm upgrade apply` and `kubeadm upgrade node`  will use `/var/lib/kubelet/instace-config.yaml`.

Reviewer: typo in instance

**GA(v1.34)**:

* `kubeadm upgrade apply` and `kubeadm upgrade node`  will use `/var/lib/kubelet/instace-config.yaml` and remove CRI annotation from Node object.

### Test Plan

> [x ] I/we understand the owners of the involved components may require updates to

Reviewer: ```suggestion
[x] I/we understand the owners of the involved components may require updates to
```

### Graduation Criteria

#### Alpha

- The upgrade will still respect the CRI annotation.
- Use `NodeLocalCRISocket` feature gate to implement features.
- Add corresponding e2e tests.

#### Beta

* Make feature gate to be enabled by default.

- Use instance kubelet configuration, ignore CRI annotations.
- Gather feedback from developers and surveys.

#### GA

- Remove the CRI annotation during upgrade(this may be not urgent or have to).

### Upgrade / Downgrade Strategy

**Alpha**: Users can patch their `ClusterConfiguration` in the `kube-system/kubeadm-config` ConfigMap to enable the `NodeLocalCRISocket` feature gate before calling kubeadm upgrade apply, which will allow a `/var/lib/kubelet/instace-config.yaml` to be generated and overwrite the `ContainerRuntimeEndpoint` field in `/var/lib/kubelet/config.yaml` with it.

**Beta**: Users can modify `ClusterConfiguration` to disable the feature gate during upgrades. This will allow them to continue using the CRI socket annotation on nodes.

**GA**: Users can no longer patch ClusterConfiguration to opt out of the feature and it will be locked to be enabled by default.

### Version Skew Strategy

> kubeadm will continue to skew from kubelet for three versions. The `ContainerRuntimeEndpoint` field in `KubeletConfiguration` was [introduced](https://github.com/kubernetes/kubernetes/pull/112136) in v1.27, so when we overwrite the `ContainerRuntimeEndpoint` field in `/var/lib/kubelet/config.yaml` through `/var/lib/kubelet/instace-config.yaml`, it will not have any impact on kubelet.

Reviewer: it will not have any impact on kubelet. -> it will be supported on all versions of the kubelet within the skew.

> ### Feature Enablement and Rollback

Reviewer: remove this and all lines before `## Implementation History`

No

### Monitoring Requirements

* N/A

### Dependencies

###### Does this feature depend on any specific services running in the cluster?

### Scalability

###### Will enabling / using this feature result in any new API calls?

No

###### Will enabling / using this feature result in introducing new API types?

No

###### Will enabling / using this feature result in any new calls to the cloud provider?

No

###### Will enabling / using this feature result in increasing size or count of the existing API objects?

No

###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?

No

###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?

No

###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?

No

### Troubleshooting

###### How does this feature react if the API server and/or etcd is unavailable?

###### What are other known failure modes?

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

> - \- 2024-05-23: Initial draft KEP

Reviewer: ```suggestion
- 2024-05-23: Initial draft KEP
```

> ## Drawbacks

Reviewer: why doing this KEP work is a bad idea.

i will give you a hint, users may read/write the alpha annotation on the Node object or the kubeadm-flags.env file and they depend on it.
this is a disruptive change.
